docs/docs/concepts/Sync_Async.md

```diff
@@ -0,0 +1,204 @@
+
+```js
+#!js api_version=1.0 name=lib
+
+redis.registerFunction('test', async function(client){
+    return client.block(function(redis_client){
+        return redis_client.call('ping');
+    });
+});
+```
+
+Running this function will return a `pong` reply:
+
+```bash
+127.0.0.1:6379> TFCALLASYNC lib.test 0
+"PONG"
+```
+
+Notice that this time, in order to invoke the function, we used `TFCALLASYNC`. **We can only invoke async functions using `TFCALLASYNC`**.
+
+Now lets look at a more complex example, assuming we want to write a function that counts the number of hashes in Redis that has name property with some value. Lets first write a synchronous function that does it, we will use the `SCAN` command to scan the key space:
+
+```js
+#!js api_version=1.0 name=lib
+
+redis.registerFunction('test', function(client, expected_name){
+    var count = 0;
+    var cursor = '0';
+    do{
+        var res = client.call('scan', cursor);
+        cursor = res[0];
+        var keys = res[1];
+        keys.forEach((key) => {
+            if (client.call('hget', key, 'name') == expected_name) {
+                count += 1;
+            }
+        });
+    } while(cursor != '0');
+    return count;
+});
+```
+
+Though working fine, this function has a potential to block Redis for a long time, lets modify this function to run on the background as a Coroutine:
+
+```js
+#!js api_version=1.0 name=lib
+
+redis.registerAsyncFunction('test', async function(async_client, expected_name){
+    var count = 0;
+    var cursor = '0';
+    do{
+        async_client.block((client)=>{
+            var res = client.call('scan', cursor);
+            cursor = res[0];
+            var keys = res[1];
+            keys.forEach((key) => {
+                if (client.call('hget', key, 'name') == expected_name) {
+                    count += 1;
+                }
+            });
+        });
+    } while(cursor != '0');
+    return count;
+});
+```
+
+Both implementations return the same result, but the seconds runs in the background and block Redis just to analyze the next batch of keys that returned from the scan command. Other commands will be processed in between the scan batches. Notice that the Coroutine approach allows the key space to be changed while the scanning it, function writer will need to decide if this is acceptable.
+
+# Start Sync and Move Async
+
+The above example is costly, even though Redis is not blocked it is still takes time to return the reply to the user. If we flatten the requirement in such way that we agree to get an approximate value, we can get a much better performance (on most cases). We will cache the result on a key called `<name>_count` and set some expiration on that key so that we will recalculate the value from time to time. The new code will look like this:
```

Comment:
```suggestion
The previous example, although functional, has a drawback in terms of performance. Even though Redis is not blocked, it still takes time to return the reply to the user. However, if we modify the requirement slightly and agree to obtain an approximate value, we can achieve much better performance in most cases. This can be done by implementing result caching using a key named `<name>_count` and setting an expiration time on that key, which triggers recalculation of the value periodically. Here's the updated code:
```